src/merkle_tree/mod.rs

```diff
@@ -1,209 +1,302 @@
+        // calculate leaf hash
+        let mut buffer = vec![0u8; leaf_size];
+        read_to_buffer(leaf, &mut buffer)?;
+        let claimed_leaf_hash = P::LeafHash::evaluate(&leaf_hash_parameters, &buffer)?;
+
+        // check if leaf corresponds to leaf hash `path`
+        if self.leaf_index & 1 == 0 && claimed_leaf_hash != self.leaf_and_sibling_hash.0 {
+            // left
+            return Ok(false);
+        } else if self.leaf_index & 1 == 1 && claimed_leaf_hash != self.leaf_and_sibling_hash.1 {
+            // right
             return Ok(false);
         }
-        // Check that the given leaf matches the leaf in the membership proof.
-        let mut buffer = [0u8; 128];
 
-        if !self.path.is_empty() {
-            let claimed_leaf_hash = hash_leaf::<P::H, L>(parameters, leaf, &mut buffer)?;
+        // check hash along the path from bottom to root
+        let mut left_buffer = vec![0u8; P::leaf_hash_output_size_upper_bound()];
+        let mut right_buffer = vec![0u8; P::leaf_hash_output_size_upper_bound()];
+        read_to_buffer(&self.leaf_and_sibling_hash.0, &mut left_buffer);
+        read_to_buffer(&self.leaf_and_sibling_hash.1, &mut right_buffer);
 
-            // Check if leaf is one of the bottom-most siblings.
-            if claimed_leaf_hash != self.path[0].0 && claimed_leaf_hash != self.path[0].1 {
-                return Ok(false);
-            };
+        let mut curr_hash =
+            P::TwoToOneHash::evaluate(&two_to_one_hash_parameters, &left_buffer, &right_buffer)?;
 
-            let mut prev = claimed_leaf_hash;
-            // Check levels between leaf level and root.
-            for &(ref hash, ref sibling_hash) in &self.path {
-                // Check if the previous hash matches the correct current hash.
-                if &prev != hash && &prev != sibling_hash {
-                    return Ok(false);
-                };
-                prev = hash_inner_node::<P::H>(parameters, hash, sibling_hash, &mut buffer)?;
-            }
+        // we will use `index` variable to track the position of path
+        let mut index = self.leaf_index;
+        index >>= 1;
 
-            if root_hash != &prev {
+        // Check levels between leaf level and root
+        for level in (0..self.non_leaf_and_sibling_hash_path.len()).rev() {
+            // check if path node at this level is left or right
+            let path_node = if index & 1 == 0 {
+                &self.non_leaf_and_sibling_hash_path[level].0
+            } else {
+                &self.non_leaf_and_sibling_hash_path[level].1
+            };
+            // if the hash calculated from last level is different from path_node, then reject
+            if &curr_hash != path_node {
                 return Ok(false);
             }
-            Ok(true)
-        } else {
-            Ok(false)
+            // update `currHash`
+            set_zero(&mut left_buffer);
+            set_zero(&mut right_buffer);
+            read_to_buffer(
```

Comment:
I'm a bit confused, why do we need set_zero and read to buffers here?

Reply:
A buffer might be needed, because we need to convert leaf or inner node to bytes. Probably we don't need `set_zero` because the length of generated byte sequence is always the same (so we can guarantee all bytes in last read are being overwritten)?

Reply:
Or do you think allocating a `vec` for each `evaluate` will not compromise the runtime? If so, I believe assigning a variable will be the best method

Reply:
Does the hash mutate its input? If not, I don't get why it has to be copied

Reply:
CRH needs to take &[u8] as input, and I believe we need to use buffer convert `&impl ToBytes` to `&[u8]`